util/fipstools/acvp/modulewrapper/modulewrapper.cc

```diff
@@ -1567,16 +1568,27 @@ static bool ECDSASigGen(const Span<const uint8_t> args[], ReplyCallback write_re
   bssl::UniquePtr<EC_KEY> key = ECKeyFromName(args[0]);
   bssl::UniquePtr<BIGNUM> d = BytesToBIGNUM(args[1]);
   const EVP_MD *hash = HashFromName(args[2]);
-  uint8_t digest[EVP_MAX_MD_SIZE];
-  unsigned digest_len;
-  if (!key || !hash ||
-      !EVP_Digest(args[3].data(), args[3].size(), digest, &digest_len, hash,
-                  /*impl=*/nullptr) ||
-      !EC_KEY_set_private_key(key.get(), d.get())) {
+  auto msg = args[3];
+  if (!key || !hash || !EC_KEY_set_private_key(key.get(), d.get())) {
     return false;
   }
-
-  bssl::UniquePtr<ECDSA_SIG> sig(ECDSA_do_sign(digest, digest_len, key.get()));
+  bssl::ScopedEVP_MD_CTX ctx;
+  EVP_PKEY_CTX *pctx;
+  bssl::UniquePtr<EVP_PKEY> evp_pkey(EVP_PKEY_new());
+  if (!evp_pkey || !EVP_PKEY_set1_EC_KEY(evp_pkey.get(), key.get())) {
+    return false;
+  }
+  std::vector<uint8_t> sig_der;
+  size_t len;
+  if (!EVP_DigestSignInit(ctx.get(), &pctx, hash, nullptr, evp_pkey.get()) ||
```

Comment:
I'll pay closer attention to the arguments, but from an initial observation, we're testing EVP_DigestSign. We're doing EVP_DigestSignInit, but not following it with Update and Final. May be we can call Init, Update, Final and compare the output with the one shot Sign. Similarly for Verify.

Reply:
`EVP_DigestSign` wraps the Update and Final.

```
if (uses_prehash(ctx, evp_sign)) {
    // If |out_sig| is NULL, the caller is only querying the maximum output
    // length. |data| should only be incorporated in the final call.
    if (out_sig != NULL &&
        !EVP_DigestSignUpdate(ctx, data, data_len)) {
      return 0;
    }

    return EVP_DigestSignFinal(ctx, out_sig, out_sig_len);
  }
```

Reply:
Does the choice of API matter? This is the single-shot one. There is also a streaming one - which could be more restrictive I guess.

Reply:
It does not matter. The lab only has requirement on if the API performs hash on the input internally.